### ISSUE TITLE
Copy text from editor onto system clipboard already fails for medium sized selections

### ISSUE THREAD

**Issue by nomennescio** [Codewars/codewars.com#1024]
Hi,

When using the editor, I sometimes need to copy into or copy from the system clipboard.
This has worked before. Now I notice that it only works if I copy ~10 lines of text with CTRL-C, but if I select more, the system clipboard will only contain a single '-' character. Why is that?

If it is a bug, please fix. If it is an intended restriction, please explain why this restriction is necessary.

**Comment by Voileexperiments** [comment]
This is a issue of CodeMirror, so maybe you should raise the issue on their github page instead.

**Comment by kazk** [comment]
@nomennescio 
I just tried 25 lines and had no problem. You can also drag and drop a local file into the editor.

I scanned some of CodeMirror's code and `"-"` seems to be from there.
I couldn't find relevant issues reported.

What is your browser and OS? Maybe this is specific to your setup and searching their issue with those information might help finding relevant information.

**Comment by nomennescio** [comment]
@kazk
Unfortunately D&D didn't work.
OS+Browser: Win 7+Chrome 59

I checked and again and came up with some numbers
37 lines OK
44 lines OK
54 lines OK
60 lines OK
61 lines FAIL
63 lines FAIL
70 lines FAIL
79 lines FAIL
84 lines FAIL
93 lines FAIL
110 lines FAIL
126 lines FAIL

I'm not sure if line length has any impact

Size of copied text (codepoints?)
981 OK
992 OK (60 lines + 1 partial line)
997 OK
999 OK
1000 OK
1001 FAIL

It seems there's a built-in limit of 1000 characters/codepoints that will be copied onto the system clipboard.

@Voileexperiments
And CodeMirror? I don't want to be fussy, but the editor is the Codewars editor, isn't it?
If it uses a component from CodeMirror, I assume Codewars maintainers would try to take up the issue with them, or make a workaround.

As a matter of fact, I would be totally happy to use an external editor for Codewars, and upload / copy-paste my code.

Thanks for your help sofar

**Comment by kazk** [comment]
> It seems there's a built-in limit of 1000 characters/codepoints that will be copied onto the system clipboard.

I tried again with lines containing 80 characters (81 including the LF) like the following.

```js
// 00000000000000000000000000000000000000000000000000000000000000000000000000000
```

I'm able to copy and paste 30 lines of this (~2400) both ways, and drop the file containing these.
I'm also using Google Chrome 59 (59.0.3071.115), but on macOS 10.12.

---

Can you try the demo <http://codemirror.net/>?
If it works there, the issue is the version used on Codewars or how it's used.

**Comment by nomennescio** [comment]
Tried the demo, same issue in there! I did not even try copy-pasting onto the system clipboard,
but just inside the demo 'panel'. Already there I saw the issue of the copied text being replaced with a single '-'.

**Comment by kazk** [comment]
> I did not even try copy-pasting onto the system clipboard,
but just inside the demo 'panel'. Already there I saw the issue of the copied text being replaced with a single '-'.

How can you see "the issue of the copied text being replaced with a single '-'" **without** "copy-pasting onto the system clipboard"? Maybe I've been misunderstanding your issue, can you clarify?

---

Do you have any extensions installed that might interfere? Try the demo with extensions turned off or in Incognito.
I think you should open an issue on CodeMirror's repo explaining how the demo fails to copy paste with your setup. The demo is using the latest version and initialized with few options, I don't think there's anything Jake can do if that fails.

**Comment by nomennescio** [comment]
I mean, I did the copy-paste inside the demo pane. The previous issue was with copying in the editor, and pasting into an external editor. In both cases the clipboard is probably used.

You did mention that you saw the '-' in de CodeMirror code? In that case wouldn't it be inherent to CodeMirror itself?

I will try to see if the same thing happens in another browser, or with extensions disabled.

**Comment by kazk** [comment]
> You did mention that you saw the '-' in de CodeMirror code? 

[src/input/TextareaInput.js#L150-L158](https://github.com/codemirror/CodeMirror/blob/faa527ee92fca8dd9eb796bdf0e2b7007b5cf260/src/input/TextareaInput.js#L150-L158) contains the following:

```js
minimal = hasCopyEvent &&
  (range.to().line - range.from().line > 100 || (selected = cm.getSelection()).length > 1000)
let content = minimal ? "-" : selected || cm.getSelection()
```

This *might* be related to your issue because `length > 1000` matches your comment, but I don't know how that piece of code is used.

> In that case wouldn't it be inherent to CodeMirror itself?

Yeah, I think the source of the issue is in CodeMirror.
I found somewhat similar issue https://github.com/codemirror/CodeMirror/issues/2703 from 2014 which was closed because it wasn't reproducible. The issue seems to be about the contents of the editor replaced with `"-"`. The issue also talks about the `minimal = hasCopyEvent`. Maybe you can help there by explaining your situation?

I'm going to close this because I don't think there's anything @jhoffner can do.

**Comment by guiambros** [comment]
Confirmed; thanks @kazk!

I got to the same conclusion: the issue is indeed coming from CodeMirror, but the root cause is elsewhere. When you're copying text that is over 1000 characters or over 100 lines, CM checks if the parent div has an oncopy event handler (see [hasCopyEvent()](https://github.com/codemirror/CodeMirror/blob/96cd8885ad759350273c71235493f6b225590442/src/util/feature_detection.js#L70)), likely using the "-" as a placeholder and letting the oncopy event handler take care of the actual text selection and copying.

If you have a Chrome extension that filters `oncopy` events - like [DFWP](https://github.com/jswanner/DontFuckWithPaste) - you're stuck with the "-" placeholder in your clipboard.

Disabling the extension, or simply [inserting the domain to be blacklisted from DFWP](https://github.com/jswanner/DontFuckWithPaste/issues/18), solves the problem.
